fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResource.java

```diff
@@ -66,19 +69,31 @@ public String transferRequestWithLoanId(@PathParam("loanId") final Long loanId,
             @Parameter(hidden = true) final String apiRequestBodyAsJson) {
         platformUserRightsContext.isAuthenticated();
 
-        throw new NotImplementedException("Not implemented yet");
+        ExternalTransferResponseData response = null;
+        if (commandParam.equals("sale")) {
+            response = externalAssetOwnersWriteService.saleLoanByLoanId(loanId, apiRequestBodyAsJson);
+        } else if (commandParam.equals("buyback")) {
+            response = externalAssetOwnersWriteService.buyBackLoanByLoanId(loanId, apiRequestBodyAsJson);
+        }
+        return apiJsonSerializerService.serialize(response);
     }
 
     @POST
     @Path("/transfers/loans/external-id/{loanExternalId}")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String transferRequestWithLoanExternalId(@PathParam("loanExternalId") final Long loanId,
+    public String transferRequestWithLoanExternalId(@PathParam("loanExternalId") final String externalLoanId,
             @QueryParam("command") @Parameter(description = "command") final String commandParam,
             @Parameter(hidden = true) final String apiRequestBodyAsJson) {
         platformUserRightsContext.isAuthenticated();
 
-        throw new NotImplementedException("Not implemented yet");
+        ExternalTransferResponseData response = null;
+        if (commandParam.equals("sale")) {
+            response = externalAssetOwnersWriteService.saleLoanByExternalLoanId(new ExternalId(externalLoanId), apiRequestBodyAsJson);
```

Comment:
Please use the `ExternalIdFactory.produce(loanExternalIdStr)` instead of manually create ExternalId object

Reply:
it would require to move 2-3 classes from provider to core, should I do that?
makes sense to do it though, because ExternalId is a common thing

Reply:
i think we should do it yes.

Reply:
I believe I fixed it everywhere